sdk/core/core/inc/az_span_emitter.h

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_SPAN_EMITTER_H
+#define AZ_SPAN_EMITTER_H
+
+#include <az_str.h>
+#include <az_span.h>
+
+#include <_az_cfg_prefix.h>
+
+AZ_ACTION_TYPE(az_span_emitter, az_span_action)
+
+/**
+ * A span of spans of bytes.
+ */
+AZ_SPAN_TYPE(az_span_span, az_span const)
+
+/**
+ * Emits all spans from @self into @action.
+ */
+AZ_NODISCARD az_result
+az_span_span_emit(az_span_span const * const self, az_span_action const action);
+
+/**
+ * @az_span_span_emit as an action of type @az_span_emitter.
+ */
+AZ_ACTION_FUNC(az_span_span_emit, az_span_span const, az_span_emitter)
+
+/**
+ * Calculates a size of a contigous buffer to store all spans from @self.
+ */
+AZ_NODISCARD az_result az_span_emitter_size(az_span_emitter const self, size_t * const out_size);
+
+/**
+ * The function creates a temporary zero-terminated string from @emmiter in dynamic memory.
```

Comment:
I find this whole "emitter" thing rather confusing. I fee like I need to trek through 2-3 layers of macros to see what's going on. It may be worth it, but it's confusing.

Reply:
This PR is a refactoring of some earlier ideas of `push` iterator. Here's some thoughts https://github.com/Azure/azure-sdk-for-c/blob/sergey/refactoring/emitter-design.md